site/viamguides/src/pydantic-validation/pydantic-validation.md

```diff
@@ -0,0 +1,647 @@
+        LOGGER.debug(f"Using latitude: {self.latitude}")
+
+        self.longitude = float(attrs.get("longitude", -121))
+        LOGGER.debug(f"Using longitude: {self.longitude}")
+
+    async def get_readings(
+        self,
+        *,
+        extra: Optional[Mapping[str, Any]] = None,
+        timeout: Optional[float] = None,
+        **kwargs
+    ) -> Mapping[
+        str,
+        SensorReading,
+    ]:
+        # Set up the Open-Meteo API client with cache and retry on error
+        cache_session = requests_cache.CachedSession(
+          '.cache', expire_after=3600)
+        retry_session = retry(cache_session, retries=5, backoff_factor=0.2)
+        openmeteo = openmeteo_requests.Client(session=retry_session)
+
+        # The order of variables in hourly or daily is
+        # important to assign them correctly below
+        url = "https://air-quality-api.open-meteo.com/v1/air-quality"
+        params = {
+            "latitude": self.latitude,
+            "longitude": self.longitude,
+            "current": ["pm10", "pm2_5"],
+            "timezone": "America/Los_Angeles"
+        }
+        responses = openmeteo.weather_api(url, params=params)
+
+        # Process location
+        response = responses[0]
+
+        # Current values. The order of variables needs
+        # to be the same as requested.
+        current = response.Current()
+        current_pm10 = current.Variables(0).Value()
+        current_pm2_5 = current.Variables(1).Value()
+
+        LOGGER.info(current_pm2_5)
+
+        # Return a dictionary of the readings
+        return {
+            "pm2_5": current_pm2_5,
+            "pm10": current_pm10
+        }
+
+
+if __name__ == "__main__":
+    asyncio.run(Module.run_from_registry())
+```
+
+7. Create the Python virtual environment and install the project dependencies:
+
+```console
+python3 -m venv .venv && source .venv/bin/activate && pip install -r requirements.txt
+```
+
+Those were quite a few steps to get through just to set up the project, so great job sticking with it! Now onto the fun part: validation!
+
+<!-- ------------------------ -->
+## Add module configuration validation
+Duration: 5
+
+Now that there is a practical project in place, you'll explore how to improve the validation logic and configuration settings with [Pydantic](https://pydantic.dev).
+
+Focusing on the `validate_config` class method first, the `config.attributes.fields` are a mapping of [attributes](https://docs.viam.com/configure/#components) set as JSON in the configuration for each component or service.
+These fields can be checked for existence (maybe they're required) and verifying the values for those fields are of a specific type; the current implementation is relatively simple for the two fields being validated.
+However, as more configuration attributes are added and extra logic (such as ensuring values are within a certain range), this will become unweildy to maintain.
```

Comment:
```suggestion
However, as more configuration attributes are added and extra logic (such as ensuring values are within a certain range), this will become unwieldy to maintain.
```